README.md
# fanctl

Automatic PID fan control for remote IPMI on Debian with Supermicro X9DRi-LN4F+

## Why

For automatically monitoring the two-zone CPU/Peripheral fan controller on a Supermicro X9DRi-LN4F using remote IPMI. Works well from a VM running on this host.

## How

In my setup, the peripheral zone controls a 140mm fan wall mounted in front of the hotswap drive bays of an SC836 3u 16-bay chassis. The CPU zone controls the CPU fans, rear case fans, and a 120mm fan mounted on the rear of the chassis behind the HBA cards to pull air through their heatsinks.

### To deploy

- clone this repo into somewhere reasonable (I used `/opt/fanctl`)
- (optional) set up a .env file to override tuning parameters and logging so you don't have to edit the scripts to do so. read the scripts for tunable params
- (optional if you don't use ZFS) edit the `spinpid2z.service` unit and remove the dependency on `zfs.service`. I added that dependency to make sure my logs land properly.
- `sudo ln -s /opt/fanctl/spinpid2z.service /etc/systemd/system/spinpid2z.service`
- `sudo systemctl enable spinpid2z`
- `sudo systemctl start spinpid2z`

## Equipment

- **Front fan wall**: [Noctua NF-A14 IndustrialPPC-3000 PWM](https://noctua.at/en/nf-a14-industrialppc-3000-pwm) x 3, arranged three wide to blow directly into the drive bays, with a PWM splitter wire running through the front IO panel into the chassis.
- **Rear PCI-e fan**: [Noctua NF-F12 IndustrialPPC-3000 PWM](https://noctua.at/en/nf-f12-industrialppc-3000-pwm) x 1, snug fit in the expansion card area on the rear exterior of the chassis, with the cable running through a perforated expansion slot shield into the chassis.
- **Interior fan wall**: [Noctua NF-A8 PWM](https://noctua.at/en/products/fan/nf-a8-pwm) x 3
- **Rear chassis fans**: [Noctua NF-A8 PWM](https://noctua.at/en/products/fan/nf-a8-pwm) x 2
- **CPU coolers**: [Noctua NH-D9DX i4 3U](https://noctua.at/en/nh-d9dx-i4-3u) x 2. Since the SC836 is a 3U server, decided to play it safe and get a cooler explicitly rated for 3U chassis compatibility.
- All 16 bays populated

## Results

### Temperatures

At idle, with a 24℃ ambient temperature, this setup keeps average drive temperatures around the temperature set point (37C as of writing) with ~40-60% duty cycle with average workloads.

### Noise levels

According to the NIOSH SLM app on my phone, the ambient noise level of my apartment is ~34 dB(A). When measured 3 feet from the server cabinet, at the entry to my closet, I measured these figures using `spintest`:

| Combined duty cycle | Noise (in dB(A)) |
| ------------------- | ---------------- |
| 100%                | 56               |
| 90%                 | 55               |
| 80%                 | 53               |
| 70%                 | 49               |
| 60%                 | 45.7             |
| 50%                 | 42.6             |
| 40%                 | 38               |
| 30%                 | 34.5             |

Since the cabinet is in a closet with an open doorway, the noise is only audible in the adjacent room if the duty cycle is above 60%, or above 40% when standing in front of it.

## Inspirations

- Largely based on https://www.truenas.com/community/threads/fan-scripts-for-supermicro-boards-using-pid-logic.51054/page-13#post-551335, with modifications:
  - made it work under Debian (original was intended for FreeBSD)
  - portability with .env file
  - remote IPMI
  - service for systemctl
  - shellcheck and formatting
- https://www.youtube.com/embed/0UjyL6ZiMkI


CHANGELOG.md
# 2021-09-12

- Added `set -euo pipefail` to the script to improve maintainability, and fixed some resulting issues with unbound variables and the like

# 2021-06-22

- Installed a USB ambient temperature sensor on the server and integrated it into the setpoint calculation for the hard drive fan speed calculation. 
  - Set `USE_AMBIENT_TEMP` in the `.env` file to enable or disable this functionality.
  - Set `AMBIENT_COEFF` in the `.env` file to adjust the desired difference between your ambient temperature and average drive temperature setpoint. Somewhere between 12 and 16 should work, depending on how hard you want your fans to work and the typical ambient temperature of your server environment.


spintest.sh
#!/usr/bin/bash
# spintest.sh for measuring DUTY-RPM relationship of fans.
VERSION="2020-06-17"
# Run as superuser. See notes at end.

##############################################
#
#  Settings
#
##############################################

#################  LOG SETTINGS ################
# Create logfile and sends all stdout and stderr to the log, as well as to the console.
LOG=/tmp/fanctl_spintest.log # Change to your desired log location/name
exec > >(tee -i $LOG) 2>&1

#################  IPMITOOL SETTING ################
# Path to ipmitool.  If you're doing VM
# you may need to add (after first quote) the following to
# remotely execute commands.
#  -H <hostname/ip> -U <username> -P <password>
IPMITOOL="/usr/bin/ipmitool"

################## Load .env file ##################

set -a
source <(cat .env | sed -e '/^#/d;/^\s*$/d' -e "s/'/'\\\''/g" -e "s/=\(.*\)/='\1'/g")
set +a

############################################################
# function print_header
# Called after user says they are ready
############################################################

function print_header() {
    echo
    DATE=$(date +"%A, %b %d, %H:%M:%S")
    echo "$DATE"

    printf "%44s %29s \n" $SPACES "___Duty%___" "Curr_RPM____________________"

    printf "%28s %9s %5s %5s %5s %5s %5s %5s \n" "MODE" "Zone0" "Zone1" "FANA" "FAN1" "FAN2" "FAN3" "FAN4"
}

#################################################
# function read_fan_data
#################################################
function read_fan_data() {

    # Read duty cycles, convert to decimal.
    # Some boards provide inaccurate duty cycle.  In that case,
    # comment out these 4 lines and we assume it is what we set.
    #DUTY_0=$($IPMITOOL raw 0x30 0x70 0x66 0 0) # in hex with leading space
    #DUTY_0=$((0x$(echo $DUTY_0)))  # strip leading space and decimalize
    #DUTY_1=$($IPMITOOL raw 0x30 0x70 0x66 0 1)
    #DUTY_1=$((0x$(echo $DUTY_1)))

    # Read fan mode, convert to decimal, get text equivalent.
    MODE=$($IPMITOOL raw 0x30 0x45 0) # in hex with leading space
    MODE=$((0x$(echo $MODE)))         # strip leading space and decimalize
    # Text for mode
    case $MODE in
    0) MODEt="Standard" ;;
    1) MODEt="Full" ;;
    2) MODEt="Optimal" ;;
    4) MODEt="HeavyIO" ;;
    esac

    # Get reported fan speed in RPM from sensor data repository.
    # Takes the pertinent FAN line, then 3 to 5 consecutive digits
    SDR=$($IPMITOOL sdr)
    FAN1=$(echo "$SDR" | grep "FAN1" | grep -Eo '[0-9]{3,5}')
    FAN2=$(echo "$SDR" | grep "FAN2" | grep -Eo '[0-9]{3,5}')
    FAN3=$(echo "$SDR" | grep "FAN3" | grep -Eo '[0-9]{3,5}')
    FAN4=$(echo "$SDR" | grep "FAN4" | grep -Eo '[0-9]{3,5}')
    FANA=$(echo "$SDR" | grep "FANA" | grep -Eo '[0-9]{3,5}')
}

##############################################
# function print_line
#
##############################################
function print_line() {
    printf "%-23s %-8s %5s %5s %5s %5s %5s %5s %5s \n" "$COND" $MODEt "${DUTY_0:----}" "${DUTY_1:----}" "${FANA:----}" "${FAN1:----}" "${FAN2:----}" "${FAN3:----}" "${FAN4:----}"

}

#####################################################
# Main section
#####################################################
read_fan_data # Get before script changes things
MODEorig=$MODE

echo
echo "Script to determine fan RPMs associated with varying duty cycles."
echo "Duty will be varied from 100% down in steps of 10."
echo "After setting duty, we wait 5 seconds for equilibration"
echo "before reading the fans.  Some points:"
echo
echo "1. Set the log path in the script before beginning."
echo "2. There should be no fan control scripts running."
echo "3. At very high and especially low fan speeds, the BMC may take "
echo "   over and change the duty.  You can:"
echo "     - ignore it,"
echo "     - adjust fan thresholds before testing, or"
echo "     - edit the script so it starts and/or ends at different values"
echo "       (see comment before while loop for instructions)."
echo
echo "Using ipmitool at $IPMITOOL"
echo "Logging to $LOG"

# Print before test
print_header
COND="Conditions before test"
print_line

echo
echo "Are you ready to begin? (y/n)"
read START
if [ $START != "y" ]; then
    exit
fi

print_header

# Set mode to 'Full' to avoid BMC changing duty cycle
# Need to wait a tick or it may not get next command
# "echo -n" to avoid annoying newline generated in log
echo -n $($IPMITOOL raw 0x30 0x45 1 1)
sleep 1

# This loop starts at 100% duty cycle and drops by
# 10 points each step, reading fans each stop.
# To limit the low end of duty cycle test,
# edit the number after '-gt' in the following line . E.g., if you
# don't want your fans to go below 20, any number from
# 10 to 19 will stop it at 20.

i=100

while [ $i -gt 10 ]; do
    DUTY_0=$i
    DUTY_1=$i
    SPEED=$(echo "scale=0; ($DUTY_0 * 255) / 100" | bc)
    echo "Speed: $SPEED"
    echo -n $($IPMITOOL raw 0x30 0x91 0x5A 0x3 0x10 "$SPEED")
    sleep 1
    echo -n $($IPMITOOL raw 0x30 0x91 0x5a 0x3 0x11 "$SPEED")
    sleep 5
    read_fan_data
    COND="Duty cycle $i%"
    print_line
    let i=i-10
done

# Leave it in previous mode and walk away
echo -n $($IPMITOOL raw 0x30 0x45 1 $MODEorig)


spincheck.sh
#!/usr/bin/bash
# spincheck.sh, logs fan and temperature data
VERSION="2020-06-17"
# Run as superuser. See notes at end.

# Creates logfile and sends all stdout and stderr to the log,
# leaving the previous contents in place. If you want to append to existing log,
# add '-a' to the tee command.
# Change to your desired log location/name:
# LOG=/mnt/MyPool/MyDataSet/MyDirectory/spincheck.log
LOG=/tmp/fanctl_spincheck.log
exec > >(tee -i $LOG) 2>&1

SP=33.57 #  Setpoint mean drive temp (C), for information only

# Path to ipmitool.  If you're doing VM
# you may need to add (after first quote) the following to
# remotely execute commands.
#  -H <hostname/ip> -U <username> -P <password>
IPMITOOL="/usr/bin/ipmitool"

##############################################
# function get_disk_name
# Get disk name from current LINE of DEVLIST
##############################################
# The awk statement works by taking $LINE as input,
# setting '(' as a _F_ield separator and taking the second field it separates
# (ie after the separator), passing that to another awk that uses
# ',' as a separator, and taking the first field (ie before the separator).
# In other words, everything between '(' and ',' is kept.

# camcontrol output for disks on HBA seems to change  every version,
# so need 2 options to get ada/da disk name.
function get_disk_name() {
    if [[ $LINE == *"(p"* ]]; then # for (pass#,[a]da#)
        DEVID=$(echo $LINE | awk -F ',' '{print $2}' | awk -F ')' '{print$1}')
    else # for ([a]da#,pass#)
        DEVID=$(echo $LINE | awk -F '(' '{print $2}' | awk -F ',' '{print$1}')
    fi
}

############################################################
# function print_header
# Called when script starts and each quarter day
############################################################
function print_header() {
    DATE=$(date +"%A, %b %d")
    printf "\n%s \n" "$DATE"
    echo -n "          "
    while read LINE; do
        get_disk_name
        printf "%-5s" $DEVID
    done <<<"$DEVLIST" # while statement works on DEVLIST
    printf "%4s %5s %5s %3s %5s %5s %5s %5s %5s %5s %5s %-7s" "Tmax" "Tmean" "ERRc" "CPU" "FAN1" "FAN2" "FAN3" "FAN4" "FANA" "Fan%0" "Fan%1" "MODE"
}

#################################################
# function manage_data: Read, process, print data
#################################################
function manage_data() {
    Tmean=$(echo "scale=3; $Tsum / $i" | bc)
    ERRc=$(echo "scale=2; $Tmean - $SP" | bc)
    # Read duty cycle, convert to decimal.
    # May need to disable these 3 lines as some boards apparently return
    # incorrect data. In that case just assume $DUTY hasn't changed.
    DUTY0=$($IPMITOOL raw 0x30 0x70 0x66 0 0) # in hex
    DUTY0=$((0x$(echo $DUTY0)))               # strip leading space and decimate
    DUTY1=$($IPMITOOL raw 0x30 0x70 0x66 0 1) # in hex
    DUTY1=$((0x$(echo $DUTY1)))               # strip leading space and decimate
    # Read fan mode, convert to decimal.
    MODE=$($IPMITOOL raw 0x30 0x45 0) # in hex
    MODE=$((0x$(echo $MODE)))         # strip leading space and decimate
    # Text for mode
    case $MODE in
    0) MODEt="Standard" ;;
    4) MODEt="HeavyIO" ;;
    2) MODEt="Optimal" ;;
    1) MODEt="Full" ;;
    esac
    # Get reported fan speed in RPM.
    # Get reported fan speed in RPM from sensor data repository.
    # Takes the pertinent FAN line, then a number with 3 to 5
    # consecutive digits
    SDR=$($IPMITOOL sdr)
    RPM_FAN1=$(echo "$SDR" | grep "FAN1" | grep -Eo '[0-9]{3,5}')
    RPM_FAN2=$(echo "$SDR" | grep "FAN2" | grep -Eo '[0-9]{3,5}')
    RPM_FAN3=$(echo "$SDR" | grep "FAN3" | grep -Eo '[0-9]{3,5}')
    RPM_FAN4=$(echo "$SDR" | grep "FAN4" | grep -Eo '[0-9]{3,5}')
    RPM_FANA=$(echo "$SDR" | grep "FANA" | grep -Eo '[0-9]{3,5}')
    # Get    # print current Tmax, Tmean
    printf "^%-3d %5.2f" $Tmax $Tmean
}

##############################################
# function DRIVES_check
# Print time on new log line.
# Go through each drive, getting and printing
# status and temp, then call function manage_data.
##############################################
function DRIVES_check() {
    echo # start new line
    TIME=$(date "+%H:%M:%S")
    echo -n "$TIME  "
    Tmax=0
    Tsum=0 # initialize drive temps for new loop through drives
    i=0    # count number of spinning drives
    while read LINE; do
        get_disk_name
        /usr/local/sbin/smartctl -a -n standby "/dev/$DEVID" >/var/tempfile
        RETURN=$? # have to preserve return value or it changes
        BIT0=$(($RETURN & 1))
        BIT1=$(($RETURN & 2))
        if [ $BIT0 -eq 0 ]; then
            if [ $BIT1 -eq 0 ]; then
                STATUS="*" # spinning
            else           # drive found but no response, probably standby
                STATUS="_"
            fi
        else # smartctl returns 1 (00000001) for missing drive
            STATUS="?"
        fi

        TEMP=""
        # Update temperatures each drive; spinners only
        if [ "$STATUS" == "*" ]; then
            # Taking 10th space-delimited field for most SATA:
            if grep -Fq "Temperature_Celsius" /var/tempfile; then
                TEMP=$(cat /var/tempfile | grep "Temperature_Celsius" | awk '{print $10}')
            # Else assume SAS, their output is:
            #     Transport protocol: SAS (SPL-3) . . .
            #     Current Drive Temperature: 45 C
            else
                TEMP=$(cat /var/tempfile | grep "Drive Temperature" | awk '{print $4}')
            fi
            let "Tsum += $TEMP"
            if [[ $TEMP > $Tmax ]]; then Tmax=$TEMP; fi
            let "i += 1"
        fi
        printf "%s%-2d  " "$STATUS" $TEMP
    done <<<"$DEVLIST"
    manage_data # manage data function
}

#####################################################
# All this happens only at the beginning
# Initializing values, list of drives, print header
#####################################################

# Check if CPU Temp is available via sysctl (will likely fail in a VM)
CPU_TEMP_SYSCTL=$(($(sysctl -a | grep dev.cpu.0.temperature | wc -l) > 0))
if [[ $CPU_TEMP_SYSCTL == 1 ]]; then
    CORES=$(($(sysctl -n hw.ncpu) - 1))
fi

echo "How many whole minutes do you want between spin checks?"
read T
SEC=$(bc <<<"$T*60") # bc is a calculator
# Get list of drives
DEVLIST1=$(/sbin/camcontrol devlist)
# Remove lines with non-spinning devices; edit as needed
# You could use another strategy, e.g., find something in the camcontrol devlist
# output that is unique to the drives you want, for instance only WDC drives:
# if [[ $LINE != *"WDC"* ]] . . .
DEVLIST="$(echo "$DEVLIST1" | sed '/KINGSTON/d;/ADATA/d;/SanDisk/d;/OCZ/d;/LSI/d;/EXP/d;/INTEL/d;/TDKMedia/d;/SSD/d;/VMware/d;/Enclosure/d;/Card/d;/Flash/d')"
DEVCOUNT=$(echo "$DEVLIST" | wc -l)
printf "\n%s\n%s\n%s\n" "NOTE ABOUT DUTY CYCLE (Fan%0 and Fan%1):" \
    "Some boards apparently report incorrect duty cycle, and can" \
    "report duty cycle for zone 1 when that zone does not exist."

# Before starting, go through the drives to report if
# smartctl return value indicates a problem (>2).
# Use -a so that all return values are available.
while read LINE; do
    get_disk_name
    /usr/local/sbin/smartctl -a -n standby "/dev/$DEVID" >/var/tempfile
    if [ $? -gt 2 ]; then
        printf "\n"
        printf "*******************************************************\n"
        printf "* WARNING - Drive %-4s has a record of past errors,   *\n" $DEVID
        printf "* is currently failing, or is not communicating well. *\n"
        printf "* Use smartctl to examine the condition of this drive *\n"
        printf "* and conduct tests. Status symbol for the drive may  *\n"
        printf "* be incorrect (but probably not).                    *\n"
        printf "*******************************************************\n"
    fi
done <<<"$DEVLIST"

printf "\n%s %36s %s \n" "Key to drive status symbols:  * spinning;  _ standby;  ? unknown" "Version" $VERSION
print_header

###########################################
# Main loop through drives every T minutes
###########################################
while [ 1 ]; do
    # Print header every quarter day.  Expression removes any
    # leading 0 so it is not seen as octal
    HM=$(date +%k%M)
    HM=$(expr $HM + 0)
    R=$((HM % 600)) # remainder after dividing by 6 hours
    if (($R < $T)); then print_header; fi
    Tmax=0
    Tsum=0 # initialize drive temps for new loop through drives
    DRIVES_check

    if [[ $CPU_TEMP_SYSCTL == 1 ]]; then
        # Find hottest CPU core
        MAX_CORE_TEMP=0
        for CORE in $(seq 0 $CORES); do
            CORE_TEMP="$(sysctl -n dev.cpu.${CORE}.temperature | awk -F '.' '{print$1}')"
            if [[ $CORE_TEMP -gt $MAX_CORE_TEMP ]]; then MAX_CORE_TEMP=$CORE_TEMP; fi
        done
        CPU_TEMP=$MAX_CORE_TEMP
    else
        CPU_TEMP=$($IPMITOOL sensor get "CPU Temp" | awk '/Sensor Reading/ {print $4}')
    fi

    # Print data.  If a fan doesn't exist, RPM value will be null.  These expressions
    # substitute a value "---" if null so printing is not messed up.  Duty cycle may be
    # reported incorrectly by boards and they can report duty for zone 1 even if there
    # is no such zone.
    printf "%6.2f %3d %5s %5s %5s %5s %5s %5d %5d %-7s" $ERRc $CPU_TEMP "${RPM_FAN1:----}" "${RPM_FAN2:----}" "${RPM_FAN3:----}" "${RPM_FAN4:----}" "${RPM_FANA:----}" $DUTY0 $DUTY1 $MODEt

    sleep $(($T * 60)) # seconds between runs
done

# Logs:
#   - disk status (spinning or standby)
#   - disk temperature (Celsius) if spinning
#   - max and mean disk temperature
#   - current 'error' of Tmean from setpoint (for information only)
#   - CPU temperature
#   - RPM for FAN1-4 and FANA
#   - duty cycle for fan zones 0 and 1
#   - fan mode

# Includes disks on motherboard and on HBA.
# Uses joeschmuck's smartctl method (returns 0 if spinning, 2 in standby)
# https://forums.freenas.org/index.php?threads/how-to-find-out-if-a-drive-is-spinning-down-properly.2068/#post-28451
# Other method (camcontrol cmd -a) doesn't work with HBA
